### PR TITLE
Improve UI scaling on iOS devices

### LINKED ISSUE #7012: UI scale on iOS is too small by default
**Attention**: I intend to work on this issue, preferably with feedback from others.

The following Screenshot is taken on mobile, on a clean build of Osu!Lazer for iOS:
78165167_3308028942602917_18363759739273216_n

I'm thinking of making a new design, maybe a skin is enough? Where buttons are much larger, as esp. the buttons on the top nav bar are small enough to be a difficult touch target.

Other UX stuff includes difficulties entering the menu inside a game, but that's not 100% related to this.

**Note**: I looked for similar issues, I could comment and work on, but couldn't find any

### PR DESCRIPTION
**Pull request by frenzibyte**
- Closes https://github.com/ppy/osu/issues/7012

Been playing around with this on my iPhone 12 and iOS simulators, and applying a general 2x scale makes the game looks pretty fitting on all iOS devices according to the simulator.

Before:
CleanShot 2022-12-24 at 00 01 51

After:
CleanShot 2022-12-23 at 23 52 04

### CODE REVIEW

**Comment by peppy** [comment]
This won't work as a blanket change like this. It should probably be based on the DPI scale of the screen, not the device flag.

Here's how it looks on an iPad:

2B99920F-A5FB-472F-AEBA-2E30E743C6B8

**Comment by bdach** [comment]
Aspect ratio of the device will also play a factor here. The root scaling container of the game is assuming 16:9, while the iPad shown above seems closer to 4:3, causing the sides to leave the screen.

**Comment by frenzibyte** [comment]
In effort of potentially pushing osu! to the app store, I've came back to this PR and experimented with it for a bit. TL;DR: Having osu! scale based on the screen resolution doesn't really work well on iOS since it's not based on a resizable window like desktop platforms. Instead, applying a general constant scale works mostly better on all iOS devices (@peppy check how it feels on your iPad now, the default scale should be slightly smaller than on master).

I've also covered Android here, but I don't have a physical device myself to tell if it looks big or not (and emulators aren't really helping, since there's no way to determine the physical size), checking on a phone and a tablet and potentially tuning the constant scale there would be appreciated.

Everything looks good for the most part, but the beatmap details area in song select need relayout or scroll support because, on my iPhone, the rating spread is cut off and the points of failure is completely masked below the screen. I have no earthly idea of how to make it look better. Ideas would be appreciated.

File: osu.Game/Graphics/Containers/ScalingContainer.cs [hunk 1]
```diff
@@ -144,6 +145,10 @@ protected override void LoadComplete()
         {
             base.LoadComplete();
 
+            // for mobile platforms, enforcing 2x scale globally fits well rather than using 1024x768 as target size.
+            if (RuntimeInfo.IsMobile)
+                content.TargetDrawSize = targetMode == ScalingMode.Everything ? DrawSize / 2f : DrawSize;
```
Comment (peppy) [inline osu.Game/Graphics/Containers/ScalingContainer.cs:150]: What happens if `targetMode` changes?

Comment (frenzibyte) [reply osu.Game/Graphics/Containers/ScalingContainer.cs:150]: `targetMode` remains constant across all containers, `scalingMode` is what the user can change and that doesn't affect this code. Querying the host window size would probably make it more readable.

Comment (frenzibyte) [reply osu.Game/Graphics/Containers/ScalingContainer.cs]: I've moved code around to hopefully make more sense, and I've also removed Android for now as I'm not confident the scale is good at all.

Comment (peppy) [reply osu.Game/Graphics/Containers/ScalingContainer.cs]: > I've also removed Android for now as I'm not confident the scale is good at all.

I don't think what you're proposing could be worse than `master`. If anything, android devices (unnecessarily) run at (unnecessarily) high pixel densities and would require more tweaking in the upwards direction.

Comment (frenzibyte) [reply osu.Game/Graphics/Containers/ScalingContainer.cs]: The part that I'm confident on is whether my changes won't make the game look too big on Android devices, but I'll push it anyway then.

**Comment by peppy** [comment]
@frenzibyte Hoping to revisit this and get it in. Could you please rebase or do something about the weird changes in the diff? 😅

File: osu.Game/OsuGameBase.cs [hunk 2]
```diff
@@ -528,7 +531,12 @@ public bool Migrate(string path)
 
         protected virtual BatteryInfo CreateBatteryInfo() => null;
 
-        protected virtual Container CreateScalingContainer() => new DrawSizePreservingFillContainer();
+        /// <summary>
+        /// The size at which all game elements would be laid out at, then scaled to the window resolution using the scaling container specified in <see cref="CreateScalingContainer"/>.
+        /// </summary>
+        protected internal virtual Vector2 TargetDrawSize => new Vector2(1024, 768);
+
+        protected virtual Container CreateScalingContainer() => new DrawSizePreservingFillContainer { TargetDrawSize = TargetDrawSize };
```
Comment (peppy) [inline osu.Game/OsuGameBase.cs:539]: This feels *so* round-about. You're seeing the value here but also reverse setting it using DI on every `Update`? And on iOS the value *is from the scaling container itself*??

I just don't know.

**Comment by peppy** [comment]
Taking a step back here, I'm pretty sure you don't want this kind of complex logic. Instead, you'd be looking to find a static value to replace `1024x768` that works better for mobile. Or failing that, rather than adding complex logic to only the iOS implementation, I think it would be better to change the way scaling works across *all* platforms.

Like, there's no reason iOS should get its own exclusive behaviour. Desktop devices could also be similar DPI / screen ratios and should be able to display just as well with default settings.

I'm in favour of spending some more time on this until it makes sense visually and code-wise.

**Comment by frenzibyte** [comment]
In that case I'll start off by closing this PR because there was no sane way I have figured out from the amount of time I've spent on this across multiple simulated devices. There's no static value that could work on all devices, and multiplying/dividing them by the device scale hasn't shown any improvements either. 

The direction this PR was heading (as I've attempted to throughly explain in https://github.com/ppy/osu/pull/21335#issuecomment-1364366546, not sure if you read that) is basically turning off size-based scaling (`TargetDrawSize` stuff) on iOS and applying a constant 2x scale to all components of the game instead, which felt pretty good to me as per the screenshots in https://github.com/ppy/osu/pull/21335#issue-1456964765, but the only approach I came up with reads rather horrifying now that I look at it again.

If you think the general direction sounds sane, then I *might* look into improving it later, but definitely don't count on me.

**Comment by peppy** [comment]
I think what makes the most sense is that we do not increase the *size* of UI elements after a certain resolution is hit, even on desktop. Whether that resolution be 1024x768 or something slightly higher would need some experimentation (on desktop) but I think it's a better approach, especially since we have "UI scale" to allow users to further tweak things to their liking.

Which is to say, this is basically what you did in this PR, but I'd see it being done using a more sane method. I think it may actually be already supported via `DrawSizePreserving` stuff.

**Pull request closed without merging**